runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5253,15 +5258,30 @@ class INTERPRETER_CLASS
 		_currentThread->callOutCount += 1;
 #endif /* JAVA_SPEC_VERSION >= 19 */
 		updateVMStruct(REGISTER_ARGS);
+#if JAVA_SPEC_VERSION >= 21
+		/* Remove VMaccess in the trivial downcall to benefit the JIT for better performance. */
+		if (!_currentThread->isInTrivialDownCall) {
+			VM_VMAccess::inlineExitVMToJNI(_currentThread);
+		}
+#else /* JAVA_SPEC_VERSION >= 21 */
 		VM_VMAccess::inlineExitVMToJNI(_currentThread);
+#endif /* JAVA_SPEC_VERSION >= 21 */
 		VM_VMHelpers::beforeJNICall(_currentThread);
 #if FFI_NATIVE_RAW_API
 		ffiCallWithSetJmpForUpcall(_currentThread, cif, function, returnStorage, values, values_raw);
 #else /* FFI_NATIVE_RAW_API */
 		ffiCallWithSetJmpForUpcall(_currentThread, cif, function, returnStorage, values);
 #endif /* FFI_NATIVE_RAW_API */
 		VM_VMHelpers::afterJNICall(_currentThread);
+#if JAVA_SPEC_VERSION >= 21
+		/* Remove VMaccess in the trivial downcall to benefit the JIT for better performance. */
+		if (!_currentThread->isInTrivialDownCall) {
+			VM_VMAccess::inlineEnterVMFromJNI(_currentThread);
+			_currentThread->isInTrivialDownCall = FALSE;
```

Comment:
This line does nothing useful (the field is already false; see line 5278.

Reply:
Fixed by moving out of the if block to ensure `isInTrivialDownCall` is always reset to false after performing the downcall.